In [1]:
import pandas as pd
import re

In [2]:
df = pd.read_csv('2024-05-05 H1N1 Dynamics.csv')
df

Dataset                                 Virus          Subject  Age  \
0      2016 UGA          H1N1 A/South Carolina/1/1918   ID_001_UGA2016   29   
1      2016 UGA                 H1N1 A/Weiss/JY2/1943   ID_001_UGA2016   29   
2      2016 UGA           H1N1 A/Fort Monmouth/1/1947   ID_001_UGA2016   29   
3      2016 UGA                  H1N1 A/Denver/1/1957   ID_001_UGA2016   29   
4      2016 UGA              H1N1 A/New Jersey/8/1976   ID_001_UGA2016   29   
...         ...                                   ...              ...  ...   
36056  2021 UGA             H1N1 A/Victoria/2570/2019  ID_252C_UGA2021   14   
36057  2021 UGA              H1N1 A/California/7/2009  ID_252C_UGA2021   14   
36058  2021 UGA                H1N1 A/Brisbane/2/2018  ID_252C_UGA2021   14   
36059  2021 UGA  H1N1 A/Guangdong-Maonan/SWL1536/2019  ID_252C_UGA2021   14   
36060  2021 UGA             H1N1 A/Victoria/2570/2019  ID_252C_UGA2021   14   

        Time  Measurement  
0       Day0            2  
1       Day0            4  
2       Day0            1  
3       Day0            0  
4       Day0            0  
...      ...          ...  
36056   Day0            2  
36057  Day33            9  
36058  Day33            8  
36059  Day33            5  
36060  Day33            7  

[36061 rows x 6 columns]

At this point, there are 36,061 rows of data. The first step is to combine data that is related to a subject so that the trends can be observed per subject. The way to do that is to isolate subjects based on the number in between their IDs, which will be referred to as a base ID. Although the IDs are going to be different for every study, subject IDs that share the same base IDs will be treated as the same subject because they are the same subject in different years.

In [6]:
def extract_base_id(subject_id):
    match = re.match(r"ID_(\w+)_UGA\d{4}", subject_id)
    return match.group(1) if match else subject_id

In [8]:
df['base_id'] = df['Subject'].apply(extract_base_id)
df

Dataset                                 Virus          Subject  Age  \
0      2016 UGA          H1N1 A/South Carolina/1/1918   ID_001_UGA2016   29   
1      2016 UGA                 H1N1 A/Weiss/JY2/1943   ID_001_UGA2016   29   
2      2016 UGA           H1N1 A/Fort Monmouth/1/1947   ID_001_UGA2016   29   
3      2016 UGA                  H1N1 A/Denver/1/1957   ID_001_UGA2016   29   
4      2016 UGA              H1N1 A/New Jersey/8/1976   ID_001_UGA2016   29   
...         ...                                   ...              ...  ...   
36056  2021 UGA             H1N1 A/Victoria/2570/2019  ID_252C_UGA2021   14   
36057  2021 UGA              H1N1 A/California/7/2009  ID_252C_UGA2021   14   
36058  2021 UGA                H1N1 A/Brisbane/2/2018  ID_252C_UGA2021   14   
36059  2021 UGA  H1N1 A/Guangdong-Maonan/SWL1536/2019  ID_252C_UGA2021   14   
36060  2021 UGA             H1N1 A/Victoria/2570/2019  ID_252C_UGA2021   14   

        Time  Measurement base_id  
0       Day0            2     001  
1       Day0            4     001  
2       Day0            1     001  
3       Day0            0     001  
4       Day0            0     001  
...      ...          ...     ...  
36056   Day0            2    252C  
36057  Day33            9    252C  
36058  Day33            8    252C  
36059  Day33            5    252C  
36060  Day33            7    252C  

[36061 rows x 7 columns]

Now that there is a base ID section in each row, this base ID will be used to create a disctionary in each row that contains a key of the viruses that the subject was vaccinated against. Inside that key's value, there will be a key of the dataset and the vaccine response measurements as the value.

In [11]:
def create_virus_dict(group):
    virus_dict = {}
    for virus, virus_group in group.groupby('Virus'):
        year_dict = {}
        for year, year_group in virus_group.groupby('Dataset'):
            year_dict[year] = year_group.sort_values('Time')['Measurement'].tolist()
        virus_dict[virus] = year_dict
    return virus_dict

In [13]:
result = df.groupby('base_id').apply(create_virus_dict).reset_index(name='Virus_Year_Measurements')

C:\Users\anall\AppData\Local\Temp\ipykernel_7336\194538327.py:1: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  result = df.groupby('base_id').apply(create_virus_dict).reset_index(name='Virus_Year_Measurements')


In [14]:
result

base_id                            Virus_Year_Measurements
0       001  {'H1N1 A/Beijing/262/1995': {'2016 UGA': [3, 3...
1      001C  {'H1N1 A/Beijing/262/1995': {'2017 UGA': [0, 2...
2       002  {'H1N1 A/Beijing/262/1995': {'2016 UGA': [4, 3...
3       003  {'H1N1 A/Beijing/262/1995': {'2016 UGA': [4, 3...
4       004  {'H1N1 A/Beijing/262/1995': {'2016 UGA': [1, 3...
..      ...                                                ...
739     560  {'H1N1 A/Brisbane/2/2018': {'2021 UGA': [4, 5]...
740     561  {'H1N1 A/Brisbane/2/2018': {'2021 UGA': [6, 6,...
741     562  {'H1N1 A/Brisbane/2/2018': {'2021 UGA': [4, 5]...
742     563  {'H1N1 A/Brisbane/2/2018': {'2021 UGA': [3, 4]...
743     564  {'H1N1 A/Brisbane/2/2018': {'2021 UGA': [4, 5]...

[744 rows x 2 columns]

Now the improvement in vaccine response measurements will be placed in an array. The value of each element represents the change in the improvement in the first and the last element of the array of vaccine measurements associated with each virus. The first element is the Day 0 measurement and the last element is the measurement taken on the final day of that test in that study.

In [18]:
improvement_arr = []
for i in range(len(result['Virus_Year_Measurements'])):
    temp_improvement_arr = []
    for virus in result['Virus_Year_Measurements'][i]:
        for dataset in result['Virus_Year_Measurements'][i][virus]:
            delta = result['Virus_Year_Measurements'][i][virus][dataset]
            temp_improvement_arr.append(delta[len(delta)-1] - delta[0])
    improvement_arr.append(temp_improvement_arr)

In [20]:
result['Improvement'] = improvement_arr
result

base_id                            Virus_Year_Measurements  \
0       001  {'H1N1 A/Beijing/262/1995': {'2016 UGA': [3, 3...   
1      001C  {'H1N1 A/Beijing/262/1995': {'2017 UGA': [0, 2...   
2       002  {'H1N1 A/Beijing/262/1995': {'2016 UGA': [4, 3...   
3       003  {'H1N1 A/Beijing/262/1995': {'2016 UGA': [4, 3...   
4       004  {'H1N1 A/Beijing/262/1995': {'2016 UGA': [1, 3...   
..      ...                                                ...   
739     560  {'H1N1 A/Brisbane/2/2018': {'2021 UGA': [4, 5]...   
740     561  {'H1N1 A/Brisbane/2/2018': {'2021 UGA': [6, 6,...   
741     562  {'H1N1 A/Brisbane/2/2018': {'2021 UGA': [4, 5]...   
742     563  {'H1N1 A/Brisbane/2/2018': {'2021 UGA': [3, 4]...   
743     564  {'H1N1 A/Brisbane/2/2018': {'2021 UGA': [4, 5]...   

                                           Improvement  
0    [0, 0, 0, 0, 4, 2, 5, 1, 0, 0, 4, 1, 1, 0, 1, ...  
1    [0, 1, 0, 0, 0, -1, 2, -1, 0, 0, 1, 1, 2, -1, ...  
2    [-1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, -1, 0, 0, 0...  
3    [-1, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 3, 1, 1]  
4     [2, 0, 2, 4, 0, 1, 0, 4, 2, 2, 2, 2, 3, 0, 0, 4]  
..                                                 ...  
739                                       [1, 1, 1, 3]  
740                                       [0, 1, 0, 1]  
741                                       [1, 2, 1, 2]  
742                                       [1, 1, 1, 1]  
743                                       [1, 2, 1, 3]  

[744 rows x 3 columns]

Now the average improvement of each subject will be calculated by taking the average of the improvement array.

In [23]:
average_arr = []
for i in range(len(result['Improvement'])):
   average_arr.append(sum(result['Improvement'][i])/len(result['Improvement'][i]))

In [25]:
result['Average Improvement'] = average_arr
result

base_id                            Virus_Year_Measurements  \
0       001  {'H1N1 A/Beijing/262/1995': {'2016 UGA': [3, 3...   
1      001C  {'H1N1 A/Beijing/262/1995': {'2017 UGA': [0, 2...   
2       002  {'H1N1 A/Beijing/262/1995': {'2016 UGA': [4, 3...   
3       003  {'H1N1 A/Beijing/262/1995': {'2016 UGA': [4, 3...   
4       004  {'H1N1 A/Beijing/262/1995': {'2016 UGA': [1, 3...   
..      ...                                                ...   
739     560  {'H1N1 A/Brisbane/2/2018': {'2021 UGA': [4, 5]...   
740     561  {'H1N1 A/Brisbane/2/2018': {'2021 UGA': [6, 6,...   
741     562  {'H1N1 A/Brisbane/2/2018': {'2021 UGA': [4, 5]...   
742     563  {'H1N1 A/Brisbane/2/2018': {'2021 UGA': [3, 4]...   
743     564  {'H1N1 A/Brisbane/2/2018': {'2021 UGA': [4, 5]...   

                                           Improvement  Average Improvement  
0    [0, 0, 0, 0, 4, 2, 5, 1, 0, 0, 4, 1, 1, 0, 1, ...              1.16129  
1    [0, 1, 0, 0, 0, -1, 2, -1, 0, 0, 1, 1, 2, -1, ...              0.36000  
2    [-1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, -1, 0, 0, 0...             -0.06250  
3    [-1, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 3, 1, 1]              0.37500  
4     [2, 0, 2, 4, 0, 1, 0, 4, 2, 2, 2, 2, 3, 0, 0, 4]              1.75000  
..                                                 ...                  ...  
739                                       [1, 1, 1, 3]              1.50000  
740                                       [0, 1, 0, 1]              0.50000  
741                                       [1, 2, 1, 2]              1.50000  
742                                       [1, 1, 1, 1]              1.00000  
743                                       [1, 2, 1, 3]              1.75000  

[744 rows x 4 columns]

In [27]:
improvement_ct = 0
for i in range(len(result)):
   if result['Average Improvement'][i] > 0:
       improvement_ct += 1

In [29]:
improvement_ct/len(result)

0.7983870967741935

Approximately 79.839% of all patients have shown improvement. Now we will split them into two groups: One-year subjects and Multi-year subjects

In order to confirm that a subject's vaccination response was studied only for a year, the subject's age and the associated dataset will be displayed on the table to ensure that the subject's ID corresponds to one study or multiple studies.

In [33]:
age_set = df.groupby('base_id').agg({'Age':set}).reset_index()

In [35]:
datasets = df.groupby('base_id').agg({'Dataset':set}).reset_index()

In [37]:
age_set_list = []
dataset_list = []
for i in range(len(age_set['Age'])):
    age_set_list.append(age_set['Age'][i])
for i in range(len(datasets['Dataset'])):
    dataset_list.append(datasets['Dataset'][i])

In [39]:
result['Age'] = age_set_list
result['Dataset'] = dataset_list

Now the subjects will be split into the one year and multi-year groups

In [42]:
one_yr_group = []
multi_yr_group = []
for i in range(len(result['Virus_Year_Measurements'])):
    all_lengths_one = all(len(value) == 1 for value in result['Virus_Year_Measurements'][i].values())
    if all_lengths_one:
        one_yr_group.append(result.loc[i])
    else:
        multi_yr_group.append(result.loc[i])

In [44]:
one_yr_group = pd.DataFrame(one_yr_group).reset_index()

In [46]:
one_yr_group

index base_id                            Virus_Year_Measurements  \
0        2     002  {'H1N1 A/Beijing/262/1995': {'2016 UGA': [4, 3...   
1        3     003  {'H1N1 A/Beijing/262/1995': {'2016 UGA': [4, 3...   
2        4     004  {'H1N1 A/Beijing/262/1995': {'2016 UGA': [1, 3...   
3        8     006  {'H1N1 A/Beijing/262/1995': {'2016 UGA': [1, 1...   
4       14     009  {'H1N1 A/Beijing/262/1995': {'2016 UGA': [4, 5...   
..     ...     ...                                                ...   
256    739     560  {'H1N1 A/Brisbane/2/2018': {'2021 UGA': [4, 5]...   
257    740     561  {'H1N1 A/Brisbane/2/2018': {'2021 UGA': [6, 6,...   
258    741     562  {'H1N1 A/Brisbane/2/2018': {'2021 UGA': [4, 5]...   
259    742     563  {'H1N1 A/Brisbane/2/2018': {'2021 UGA': [3, 4]...   
260    743     564  {'H1N1 A/Brisbane/2/2018': {'2021 UGA': [4, 5]...   

                                           Improvement  Average Improvement  \
0    [-1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, -1, 0, 0, 0...              -0.0625   
1    [-1, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 3, 1, 1]               0.3750   
2     [2, 0, 2, 4, 0, 1, 0, 4, 2, 2, 2, 2, 3, 0, 0, 4]               1.7500   
3     [0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0]               0.1875   
4    [1, 0, 1, 2, 0, 0, 0, 1, 0, 0, -1, 1, 0, -1, 1...               0.3125   
..                                                 ...                  ...   
256                                       [1, 1, 1, 3]               1.5000   
257                                       [0, 1, 0, 1]               0.5000   
258                                       [1, 2, 1, 2]               1.5000   
259                                       [1, 1, 1, 1]               1.0000   
260                                       [1, 2, 1, 3]               1.7500   

      Age     Dataset  
0    {29}  {2016 UGA}  
1    {28}  {2016 UGA}  
2    {27}  {2016 UGA}  
3    {35}  {2016 UGA}  
4    {61}  {2016 UGA}  
..    ...         ...  
256  {18}  {2021 UGA}  
257  {18}  {2021 UGA}  
258  {18}  {2021 UGA}  
259  {18}  {2021 UGA}  
260  {18}  {2021 UGA}  

[261 rows x 7 columns]

In [48]:
improvement_ct = 0
for i in range(len(one_yr_group)):
   if one_yr_group['Average Improvement'][i] > 0:
       improvement_ct += 1
improvement_ct/len(one_yr_group)

0.8467432950191571

In [50]:
dataset_one_yr = []
for i in range(len(datasets['Dataset'])):
    if len(datasets['Dataset'][i]) == 1:
        dataset_one_yr.append(datasets.loc[i])
dataset_one_yr = pd.DataFrame(dataset_one_yr).reset_index()

In [52]:
dataset_one_yr

index base_id     Dataset
0        2     002  {2016 UGA}
1        3     003  {2016 UGA}
2        4     004  {2016 UGA}
3        8     006  {2016 UGA}
4       14     009  {2016 UGA}
..     ...     ...         ...
256    739     560  {2021 UGA}
257    740     561  {2021 UGA}
258    741     562  {2021 UGA}
259    742     563  {2021 UGA}
260    743     564  {2021 UGA}

[261 rows x 3 columns]

The indices and subject base IDs match with the table before, which means that the data matches. Now the same process will be applied onto the multiple year group.

In [55]:
multi_yr_group = pd.DataFrame(multi_yr_group).reset_index()
multi_yr_group

index base_id                            Virus_Year_Measurements  \
0        0     001  {'H1N1 A/Beijing/262/1995': {'2016 UGA': [3, 3...   
1        1    001C  {'H1N1 A/Beijing/262/1995': {'2017 UGA': [0, 2...   
2        5    004C  {'H1N1 A/Beijing/262/1995': {'2017 UGA': [3, 3...   
3        6     005  {'H1N1 A/Beijing/262/1995': {'2016 UGA': [1, 0...   
4        7    005C  {'H1N1 A/Beijing/262/1995': {'2017 UGA': [2, 2...   
..     ...     ...                                                ...   
478    721     526  {'H1N1 A/Brisbane/2/2018': {'2019 UGA': [5, 4,...   
479    726     531  {'H1N1 A/Brisbane/2/2018': {'2019 UGA': [2, 2,...   
480    727     535  {'H1N1 A/Brisbane/2/2018': {'2019 UGA': [1, 6]...   
481    728     539  {'H1N1 A/Brisbane/2/2018': {'2020 UGA': [0, 5,...   
482    729     540  {'H1N1 A/Brisbane/2/2018': {'2020 UGA': [3, 5,...   

                                           Improvement  Average Improvement  \
0    [0, 0, 0, 0, 4, 2, 5, 1, 0, 0, 4, 1, 1, 0, 1, ...             1.161290   
1    [0, 1, 0, 0, 0, -1, 2, -1, 0, 0, 1, 1, 2, -1, ...             0.360000   
2    [0, 0, 0, 1, 1, 0, -1, 0, -1, -1, 1, 0, -4, 0,...             0.055556   
3    [-1, 0, 0, 0, 0, -1, 0, 2, 1, -1, 0, 0, 0, 0, ...             0.000000   
4    [0, 0, 0, 1, 0, 2, 0, -3, 0, 0, 0, 4, -1, 1, 0...             0.440000   
..                                                 ...                  ...   
478    [1, 1, 0, 0, 0, -2, 1, 0, 0, 0, 0, 0, 0, -1, 0]             0.000000   
479      [2, 0, 3, 1, 3, 0, 3, 0, 2, 2, 1, 1, 3, 0, 3]             1.600000   
480                   [5, 0, 0, 1, -1, 0, 5, -1, 4, 0]             1.300000   
481                        [5, 0, 1, 0, 4, 0, 2, 1, 0]             1.444444   
482                        [1, 2, 1, 3, 2, 2, 3, 0, 3]             1.888889   

                      Age                                            Dataset  
0        {32, 33, 29, 31}           {2016 UGA, 2019 UGA, 2021 UGA, 2020 UGA}  
1                {16, 15}                               {2017 UGA, 2018 UGA}  
2        {16, 17, 14, 15}           {2017 UGA, 2019 UGA, 2018 UGA, 2020 UGA}  
3    {25, 26, 27, 28, 29}  {2021 UGA, 2016 UGA, 2020 UGA, 2018 UGA, 2017 ...  
4                {16, 17}                               {2017 UGA, 2018 UGA}  
..                    ...                                                ...  
478              {27, 28}                     {2019 UGA, 2020 UGA, 2021 UGA}  
479          {19, 20, 21}                     {2019 UGA, 2020 UGA, 2021 UGA}  
480              {65, 63}                               {2019 UGA, 2021 UGA}  
481                  {18}                               {2020 UGA, 2021 UGA}  
482                  {18}                               {2020 UGA, 2021 UGA}  

[483 rows x 7 columns]

In [57]:
improvement_ct = 0
for i in range(len(multi_yr_group)):
   if multi_yr_group['Average Improvement'][i] > 0:
       improvement_ct += 1
improvement_ct/len(multi_yr_group)

0.772256728778468

In [59]:
dataset_multi_yr = []
for i in range(len(datasets['Dataset'])):
    if len(datasets['Dataset'][i]) != 1:
        dataset_multi_yr.append(datasets.loc[i])
dataset_multi_yr = pd.DataFrame(dataset_multi_yr)

In [61]:
dataset_multi_yr

base_id                                            Dataset
0       001           {2016 UGA, 2019 UGA, 2021 UGA, 2020 UGA}
1      001C                               {2017 UGA, 2018 UGA}
5      004C           {2017 UGA, 2019 UGA, 2018 UGA, 2020 UGA}
6       005  {2021 UGA, 2016 UGA, 2020 UGA, 2018 UGA, 2017 ...
7      005C                               {2017 UGA, 2018 UGA}
..      ...                                                ...
721     526                     {2019 UGA, 2020 UGA, 2021 UGA}
726     531                     {2019 UGA, 2020 UGA, 2021 UGA}
727     535                               {2019 UGA, 2021 UGA}
728     539                               {2020 UGA, 2021 UGA}
729     540                               {2020 UGA, 2021 UGA}

[483 rows x 2 columns]

Indices match so it is confirmed that the table of vaccine data relates to the multi-year patients.

In [64]:
new_age = []

for i in range(len(one_yr_group)):
    new_age_list = list(one_yr_group['Age'][i])
    new_age.append(new_age_list[0])

one_yr_group['Updated Age'] = new_age

Each group will be split into 6 groups as listed in the comments next to the definition of the array of each group.

In [67]:
group1_one_yr = [] # Infants(<1 year old)
group2_one_yr = [] # Children(1-11 years old)
group3_one_yr = [] # Adolescents(12-17 years old)
group4_one_yr = [] # Adults(18-44 years old)
group5_one_yr = [] # Middle Aged Adults(45-64 years old)
group6_one_yr = [] # Seniors(>65 years old)
for i in range(len(one_yr_group)):
    if one_yr_group['Updated Age'][i] < 1:
        group1_one_yr.append(one_yr_group.loc[i])
    elif one_yr_group['Updated Age'][i] < 12:
        group2_one_yr.append(one_yr_group.loc[i])
    elif one_yr_group['Updated Age'][i] < 18:
        group3_one_yr.append(one_yr_group.loc[i])
    elif one_yr_group['Updated Age'][i] < 45:
        group4_one_yr.append(one_yr_group.loc[i])
    elif one_yr_group['Updated Age'][i] < 65:
        group5_one_yr.append(one_yr_group.loc[i])
    else:
        group6_one_yr.append(one_yr_group.loc[i])

In [69]:
group1_one_yr

[]

There's no available data for infants.

In [72]:
group2_one_yr = pd.DataFrame(group2_one_yr).reset_index()
group2_one_yr
improvement_arr = []
for i in range(len(group2_one_yr['Virus_Year_Measurements'])):
    temp_improvement_arr = []
    for virus in group2_one_yr['Virus_Year_Measurements'][i]:
        for dataset in group2_one_yr['Virus_Year_Measurements'][i][virus]:
            delta = group2_one_yr['Virus_Year_Measurements'][i][virus][dataset]
            temp_improvement_arr.append(delta[len(delta)-1] - delta[0])
    improvement_arr.append(temp_improvement_arr)
group2_one_yr['Improvement'] = improvement_arr
average_arr = []
for i in range(len(group2_one_yr['Improvement'])):
   average_arr.append(sum(group2_one_yr['Improvement'][i])/len(group2_one_yr['Improvement'][i]))
group2_one_yr['Average Improvement'] = average_arr
group2_one_yr

level_0  index base_id                            Virus_Year_Measurements  \
0       118    369    206C  {'H1N1 A/Brisbane/2/2018': {'2021 UGA': [3, 6,...   
1       119    371    207C  {'H1N1 A/Brisbane/2/2018': {'2021 UGA': [7, 6,...   
2       124    382    212C  {'H1N1 A/Brisbane/2/2018': {'2021 UGA': [5, 5,...   
3       127    388    215C  {'H1N1 A/Brisbane/2/2018': {'2021 UGA': [0, 0]...   
4       129    390    216C  {'H1N1 A/Brisbane/2/2018': {'2021 UGA': [3, 8]...   
5       131    395    219C  {'H1N1 A/Brisbane/2/2018': {'2021 UGA': [3, 3]...   
6       132    397    220C  {'H1N1 A/Brisbane/2/2018': {'2021 UGA': [5, 5]...   
7       133    399    221C  {'H1N1 A/Brisbane/2/2018': {'2021 UGA': [4, 4]...   
8       136    403    223C  {'H1N1 A/Brisbane/2/2018': {'2021 UGA': [5, 7]...   
9       139    408    226C  {'H1N1 A/Brisbane/2/2018': {'2021 UGA': [5, 7]...   
10      145    416    230C  {'H1N1 A/Brisbane/2/2018': {'2021 UGA': [5, 5,...   
11      150    422    233C  {'H1N1 A/Brisbane/2/2018': {'2021 UGA': [3, 6]...   
12      154    426    235C  {'H1N1 A/Brisbane/2/2018': {'2021 UGA': [1, 5,...   
13      157    430    237C  {'H1N1 A/Brisbane/2/2018': {'2021 UGA': [5, 6]...   
14      163    438    243C  {'H1N1 A/Brisbane/2/2018': {'2021 UGA': [4, 7]...   
15      168    449    249C  {'H1N1 A/Brisbane/2/2018': {'2021 UGA': [5, 6]...   

       Improvement  Average Improvement   Age     Dataset  Updated Age  
0    [1, 2, -1, 2]                 1.00  {11}  {2021 UGA}           11  
1     [0, 1, 0, 1]                 0.50  {11}  {2021 UGA}           11  
2    [0, 1, -1, 0]                 0.00  {10}  {2021 UGA}           10  
3     [0, 2, 3, 2]                 1.75  {11}  {2021 UGA}           11  
4     [5, 4, 3, 5]                 4.25  {10}  {2021 UGA}           10  
5     [0, 0, 0, 1]                 0.25  {10}  {2021 UGA}           10  
6     [0, 0, 0, 1]                 0.25  {10}  {2021 UGA}           10  
7     [0, 1, 0, 3]                 1.00  {11}  {2021 UGA}           11  
8     [2, 1, 1, 0]                 1.00  {10}  {2021 UGA}           10  
9     [2, 2, 2, 3]                 2.25  {11}  {2021 UGA}           11  
10    [0, 1, 0, 1]                 0.50  {10}  {2021 UGA}           10  
11    [3, 5, 3, 5]                 4.00  {11}  {2021 UGA}           11  
12  [-1, 1, 0, -1]                -0.25  {11}  {2021 UGA}           11  
13    [1, 2, 2, 2]                 1.75  {11}  {2021 UGA}           11  
14    [3, 3, 2, 5]                 3.25  {10}  {2021 UGA}           10  
15    [1, 0, 0, 1]                 0.50  {10}  {2021 UGA}           10

In [74]:
avg_improvement = sum(group2_one_yr['Average Improvement'])/len(group2_one_yr)
avg_improvement

1.375

In [76]:
improvement_ct = 0
for i in range(len(group2_one_yr)):
   if group2_one_yr['Average Improvement'][i] > 0:
       improvement_ct += 1
improvement_ct/len(group2_one_yr)

0.875

The average improvement for children that have been tested for one year is 1.375.

In [79]:
group3_one_yr = pd.DataFrame(group3_one_yr).reset_index()
improvement_arr = []
for i in range(len(group3_one_yr['Virus_Year_Measurements'])):
    temp_improvement_arr = []
    for virus in group3_one_yr['Virus_Year_Measurements'][i]:
        for dataset in group3_one_yr['Virus_Year_Measurements'][i][virus]:
            delta = group3_one_yr['Virus_Year_Measurements'][i][virus][dataset]
            temp_improvement_arr.append(delta[len(delta)-1] - delta[0])
    improvement_arr.append(temp_improvement_arr)
group3_one_yr['Improvement'] = improvement_arr
average_arr = []
for i in range(len(group3_one_yr['Improvement'])):
   average_arr.append(sum(group3_one_yr['Improvement'][i])/len(group3_one_yr['Improvement'][i]))
group3_one_yr['Average Improvement'] = average_arr
group3_one_yr

level_0  index base_id                            Virus_Year_Measurements  \
0         5     17    010C  {'H1N1 A/Beijing/262/1995': {'2017 UGA': [2, 2...   
1         6     19    011C  {'H1N1 A/Beijing/262/1995': {'2017 UGA': [3, 3...   
2         8     21    012C  {'H1N1 A/Beijing/262/1995': {'2017 UGA': [2, 2...   
3        10     40    022C  {'H1N1 A/Beijing/262/1995': {'2017 UGA': [4, 4...   
4        16     60    033C  {'H1N1 A/Beijing/262/1995': {'2017 UGA': [1, 1...   
..      ...    ...     ...                                                ...   
73      166    444    246C  {'H1N1 A/Brisbane/2/2018': {'2021 UGA': [6, 6]...   
74      167    447    248C  {'H1N1 A/Brisbane/2/2018': {'2021 UGA': [2, 3]...   
75      169    451    250C  {'H1N1 A/Brisbane/2/2018': {'2021 UGA': [6, 6]...   
76      170    453    251C  {'H1N1 A/Brisbane/2/2018': {'2021 UGA': [3, 6]...   
77      171    455    252C  {'H1N1 A/Brisbane/2/2018': {'2021 UGA': [3, 8]...   

                                          Improvement  Average Improvement  \
0    [0, 0, 1, 0, 0, 0, 0, 1, 1, 0, 1, 0, 0, 0, 0, 0]               0.2500   
1   [0, 0, 0, -1, 0, 0, 0, 1, 1, 0, 1, 0, 1, 0, 0, 0]               0.1875   
2    [0, 0, 2, 3, 0, 0, 0, 2, 1, 2, 2, 0, 2, 1, 0, 0]               0.9375   
3   [-1, 0, 0, -1, -4, 1, 0, 0, 0, 3, -1, -1, 3, -...               0.0000   
4   [0, 0, 1, 5, 0, 0, 0, 6, 1, 6, 3, -2, 5, 2, 0, 0]               1.6875   
..                                                ...                  ...   
73                                       [0, 1, 0, 1]               0.5000   
74                                       [1, 2, 0, 1]               1.0000   
75                                       [0, 2, 1, 1]               1.0000   
76                                       [3, 4, 2, 3]               3.0000   
77                                       [5, 5, 3, 5]               4.5000   

     Age     Dataset  Updated Age  
0   {15}  {2017 UGA}           15  
1   {17}  {2017 UGA}           17  
2   {17}  {2017 UGA}           17  
3   {17}  {2017 UGA}           17  
4   {17}  {2017 UGA}           17  
..   ...         ...          ...  
73  {13}  {2021 UGA}           13  
74  {14}  {2021 UGA}           14  
75  {13}  {2021 UGA}           13  
76  {12}  {2021 UGA}           12  
77  {14}  {2021 UGA}           14  

[78 rows x 9 columns]

In [81]:
avg_improvement = sum(group3_one_yr['Average Improvement'])/len(group3_one_yr)
avg_improvement

1.0338319088319088

In [83]:
improvement_ct = 0
for i in range(len(group3_one_yr)):
   if group3_one_yr['Average Improvement'][i] > 0:
       improvement_ct += 1
improvement_ct/len(group3_one_yr)

0.8717948717948718

The average improvement in adolescents who were tested for 1 year is 1.03383.

In [86]:
group4_one_yr = pd.DataFrame(group4_one_yr).reset_index()
improvement_arr = []
for i in range(len(group4_one_yr['Virus_Year_Measurements'])):
    temp_improvement_arr = []
    for virus in group4_one_yr['Virus_Year_Measurements'][i]:
        for dataset in group4_one_yr['Virus_Year_Measurements'][i][virus]:
            delta = group4_one_yr['Virus_Year_Measurements'][i][virus][dataset]
            temp_improvement_arr.append(delta[len(delta)-1] - delta[0])
    improvement_arr.append(temp_improvement_arr)
group4_one_yr['Improvement'] = improvement_arr
average_arr = []
for i in range(len(group4_one_yr['Improvement'])):
   average_arr.append(sum(group4_one_yr['Improvement'][i])/len(group4_one_yr['Improvement'][i]))
group4_one_yr['Average Improvement'] = average_arr
group4_one_yr

level_0  index base_id  \
0          0      2     002   
1          1      3     003   
2          2      4     004   
3          3      8     006   
4          7     20     012   
..       ...    ...     ...   
108      256    739     560   
109      257    740     561   
110      258    741     562   
111      259    742     563   
112      260    743     564   

                               Virus_Year_Measurements  \
0    {'H1N1 A/Beijing/262/1995': {'2016 UGA': [4, 3...   
1    {'H1N1 A/Beijing/262/1995': {'2016 UGA': [4, 3...   
2    {'H1N1 A/Beijing/262/1995': {'2016 UGA': [1, 3...   
3    {'H1N1 A/Beijing/262/1995': {'2016 UGA': [1, 1...   
4    {'H1N1 A/Beijing/262/1995': {'2016 UGA': [4, 4...   
..                                                 ...   
108  {'H1N1 A/Brisbane/2/2018': {'2021 UGA': [4, 5]...   
109  {'H1N1 A/Brisbane/2/2018': {'2021 UGA': [6, 6,...   
110  {'H1N1 A/Brisbane/2/2018': {'2021 UGA': [4, 5]...   
111  {'H1N1 A/Brisbane/2/2018': {'2021 UGA': [3, 4]...   
112  {'H1N1 A/Brisbane/2/2018': {'2021 UGA': [4, 5]...   

                                           Improvement  Average Improvement  \
0    [-1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, -1, 0, 0, 0...              -0.0625   
1    [-1, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 3, 1, 1]               0.3750   
2     [2, 0, 2, 4, 0, 1, 0, 4, 2, 2, 2, 2, 3, 0, 0, 4]               1.7500   
3     [0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0]               0.1875   
4    [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, -1, 0, 0, 0, 1, 0]               0.0000   
..                                                 ...                  ...   
108                                       [1, 1, 1, 3]               1.5000   
109                                       [0, 1, 0, 1]               0.5000   
110                                       [1, 2, 1, 2]               1.5000   
111                                       [1, 1, 1, 1]               1.0000   
112                                       [1, 2, 1, 3]               1.7500   

      Age     Dataset  Updated Age  
0    {29}  {2016 UGA}           29  
1    {28}  {2016 UGA}           28  
2    {27}  {2016 UGA}           27  
3    {35}  {2016 UGA}           35  
4    {24}  {2016 UGA}           24  
..    ...         ...          ...  
108  {18}  {2021 UGA}           18  
109  {18}  {2021 UGA}           18  
110  {18}  {2021 UGA}           18  
111  {18}  {2021 UGA}           18  
112  {18}  {2021 UGA}           18  

[113 rows x 9 columns]

In [88]:
avg_improvement = sum(group4_one_yr['Average Improvement'])/len(group4_one_yr)
avg_improvement

0.9637045231071779

In [90]:
improvement_ct = 0
for i in range(len(group4_one_yr)):
   if group4_one_yr['Average Improvement'][i] > 0:
       improvement_ct += 1
improvement_ct/len(group4_one_yr)

0.8584070796460177

The average improvement in adults 18-44 who were vaccinated for one year is 0.9637.

In [93]:
group5_one_yr = pd.DataFrame(group5_one_yr).reset_index()
improvement_arr = []
for i in range(len(group5_one_yr['Virus_Year_Measurements'])):
    temp_improvement_arr = []
    for virus in group5_one_yr['Virus_Year_Measurements'][i]:
        for dataset in group5_one_yr['Virus_Year_Measurements'][i][virus]:
            delta = group5_one_yr['Virus_Year_Measurements'][i][virus][dataset]
            temp_improvement_arr.append(delta[len(delta)-1] - delta[0])
    improvement_arr.append(temp_improvement_arr)
group5_one_yr['Improvement'] = improvement_arr
average_arr = []
for i in range(len(group5_one_yr['Improvement'])):
   average_arr.append(sum(group5_one_yr['Improvement'][i])/len(group5_one_yr['Improvement'][i]))
group5_one_yr['Average Improvement'] = average_arr
group5_one_yr

level_0  index base_id                            Virus_Year_Measurements  \
0         4     14     009  {'H1N1 A/Beijing/262/1995': {'2016 UGA': [4, 5...   
1        39    147     079  {'H1N1 A/Beijing/262/1995': {'2016 UGA': [1, 1...   
2        56    183     098  {'H1N1 A/Beijing/262/1995': {'2016 UGA': [1, 1...   
3        59    190     102  {'H1N1 A/Beijing/262/1995': {'2016 UGA': [3, 3...   
4        66    210     113  {'H1N1 A/Beijing/262/1995': {'2016 UGA': [2, 2...   
5        76    243     131  {'H1N1 A/Beijing/262/1995': {'2016 UGA': [2, 0...   
6        89    300     163  {'H1N1 A/Beijing/262/1995': {'2017 UGA': [0, 1...   
7        96    318     173  {'H1N1 A/Beijing/262/1995': {'2017 UGA': [0, 1...   
8       123    381     212  {'H1N1 A/Beijing/262/1995': {'2017 UGA': [2, 3...   
9       135    402     223  {'H1N1 A/Beijing/262/1995': {'2017 UGA': [0, 2...   
10      172    458     255  {'H1N1 A/Beijing/262/1995': {'2017 UGA': [1, 4...   
11      174    460     257  {'H1N1 A/Beijing/262/1995': {'2017 UGA': [1, 0...   
12      175    462     261  {'H1N1 A/Beijing/262/1995': {'2017 UGA': [1, 0...   
13      180    473     272  {'H1N1 A/Brisbane/2/2018': {'2019 UGA': [0, 3]...   
14      186    515     315  {'H1N1 A/Brisbane/2/2018': {'2019 UGA': [0, 2]...   
15      187    524     324  {'H1N1 A/Brisbane/2/2018': {'2019 UGA': [0, 0]...   
16      188    527     327  {'H1N1 A/Brisbane/2/2018': {'2019 UGA': [0, 3]...   
17      189    541     341  {'H1N1 A/Brisbane/2/2018': {'2019 UGA': [1, 3]...   
18      193    557     357  {'H1N1 A/Brisbane/2/2018': {'2019 UGA': [0, 1]...   
19      194    564     365  {'H1N1 A/Brisbane/2/2018': {'2019 UGA': [1, 7]...   
20      195    565     366  {'H1N1 A/Brisbane/2/2018': {'2019 UGA': [4, 5]...   
21      196    589     390  {'H1N1 A/Brisbane/2/2018': {'2019 UGA': [3, 3]...   
22      197    598     400  {'H1N1 A/Brisbane/2/2018': {'2019 UGA': [2, 2]...   
23      198    600     402  {'H1N1 A/Brisbane/2/2018': {'2019 UGA': [0, 2]...   
24      201    613     415  {'H1N1 A/Brisbane/2/2018': {'2019 UGA': [1, 3]...   
25      217    666     469  {'H1N1 A/Brisbane/2/2018': {'2019 UGA': [1, 1]...   
26      221    679     482  {'H1N1 A/Brisbane/2/2018': {'2019 UGA': [0, 0]...   
27      229    698     501  {'H1N1 A/Brisbane/2/2018': {'2019 UGA': [5, 5]...   
28      230    700     503  {'H1N1 A/Brisbane/2/2018': {'2019 UGA': [0, 0]...   
29      231    701     504  {'H1N1 A/Brisbane/2/2018': {'2019 UGA': [1, 7]...   
30      237    708     511  {'H1N1 A/Brisbane/2/2018': {'2019 UGA': [3, 3]...   
31      238    710     513  {'H1N1 A/Brisbane/2/2018': {'2019 UGA': [1, 2]...   
32      239    711     514  {'H1N1 A/Brisbane/2/2018': {'2019 UGA': [5, 5]...   
33      240    713     517  {'H1N1 A/Brisbane/2/2018': {'2019 UGA': [5, 5]...   

                                          Improvement  Average Improvement  \
0   [1, 0, 1, 2, 0, 0, 0, 1, 0, 0, -1, 1, 0, -1, 1...             0.312500   
1    [0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0]             0.125000   
2   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, -5, 0]            -0.312500   
3    [0, 0, 2, 1, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0]             0.312500   
4    [0, 2, 3, 5, 3, 1, 2, 4, 1, 2, 2, 1, 3, 0, 3, 1]             2.062500   
5   [-2, -1, 0, 0, 0, 0, 0, 0, -1, -1, 0, -1, 0, 0...            -0.375000   
6   [1, 0, 0, 3, 1, 0, 0, 4, -1, 1, 1, 0, 2, 0, 0, 1]             0.812500   
7   [1, 1, -1, 1, 0, 0, 0, 2, 0, 1, -1, 1, 0, 0, 1...             0.437500   
8    [1, 1, 0, 2, 1, 0, 1, 2, 0, 1, 1, 0, 0, 0, 1, 0]             0.687500   
9   [2, -1, 0, 1, 0, 0, -1, 0, 0, 1, -1, 1, 0, 0, ...             0.125000   
10   [3, 4, 4, 5, 3, 5, 3, 6, 4, 3, 4, 4, 3, 4, 4, 1]             3.750000   
11  [-1, 0, 0, 3, 1, 1, 1, 5, 1, 4, 2, 0, 3, 1, 0, 0]             1.312500   
12  [-1, 2, 0, 2, 1, 3, 2, 4, 0, 2, 3, -2, 2, 2, 1...             1.375000   
13                                 [3, 0, 2, 0, 2, 4]             1.833333  

In [95]:
avg_improvement = sum(group5_one_yr['Average Improvement'])/len(group5_one_yr)
avg_improvement

0.8615196078431373

In [97]:
improvement_ct = 0
for i in range(len(group5_one_yr)):
   if group5_one_yr['Average Improvement'][i] > 0:
       improvement_ct += 1
improvement_ct/len(group5_one_yr)

0.7647058823529411

The average improvement in older adults 45-62 who were vaccinated for a year is 0.8615.

In [100]:
group6_one_yr = pd.DataFrame(group6_one_yr).reset_index()
improvement_arr = []
for i in range(len(group6_one_yr['Virus_Year_Measurements'])):
    temp_improvement_arr = []
    for virus in group6_one_yr['Virus_Year_Measurements'][i]:
        for dataset in group6_one_yr['Virus_Year_Measurements'][i][virus]:
            delta = group6_one_yr['Virus_Year_Measurements'][i][virus][dataset]
            temp_improvement_arr.append(delta[len(delta)-1] - delta[0])
    improvement_arr.append(temp_improvement_arr)
group6_one_yr['Improvement'] = improvement_arr
average_arr = []
for i in range(len(group6_one_yr['Improvement'])):
   average_arr.append(sum(group6_one_yr['Improvement'][i])/len(group6_one_yr['Improvement'][i]))
group6_one_yr['Average Improvement'] = average_arr
group6_one_yr

level_0  index base_id                            Virus_Year_Measurements  \
0        54    179     096  {'H1N1 A/Beijing/262/1995': {'2016 UGA': [1, 1...   
1        87    294     160  {'H1N1 A/Beijing/262/1995': {'2017 UGA': [0, 0...   
2        93    308     168  {'H1N1 A/Beijing/262/1995': {'2017 UGA': [1, 1...   
3        95    316     172  {'H1N1 A/Beijing/262/1995': {'2017 UGA': [0, 2...   
4        98    324     176  {'H1N1 A/Beijing/262/1995': {'2017 UGA': [2, 2...   
5        99    326     178  {'H1N1 A/Beijing/262/1995': {'2017 UGA': [0, 0...   
6       100    329     180  {'H1N1 A/Beijing/262/1995': {'2017 UGA': [2, 2...   
7       102    333     182  {'H1N1 A/Beijing/262/1995': {'2017 UGA': [1, 1...   
8       103    338     186  {'H1N1 A/Beijing/262/1995': {'2017 UGA': [0, 0...   
9       108    346     192  {'H1N1 A/Beijing/262/1995': {'2017 UGA': [1, 2...   
10      110    348     193  {'H1N1 A/Beijing/262/1995': {'2017 UGA': [1, 1...   
11      114    359     200  {'H1N1 A/Beijing/262/1995': {'2017 UGA': [1, 1...   
12      115    361     201  {'H1N1 A/Beijing/262/1995': {'2017 UGA': [0, 0...   
13      179    472     271  {'H1N1 A/Brisbane/2/2018': {'2019 UGA': [1, 1]...   
14      183    488     287  {'H1N1 A/Brisbane/2/2018': {'2019 UGA': [1, 2]...   
15      184    493     292  {'H1N1 A/Brisbane/2/2018': {'2019 UGA': [0, 2]...   
16      185    510     310  {'H1N1 A/Brisbane/2/2018': {'2019 UGA': [3, 5]...   
17      190    542     342  {'H1N1 A/Brisbane/2/2018': {'2019 UGA': [0, 0]...   
18      191    550     350  {'H1N1 A/Brisbane/2/2018': {'2019 UGA': [0, 1]...   
19      199    604     406  {'H1N1 A/Brisbane/2/2018': {'2019 UGA': [0, 3]...   

                                          Improvement  Average Improvement  \
0   [0, 0, 0, 3, 1, 0, 0, 3, 1, 3, 1, 2, 3, -1, 0, 0]             1.000000   
1   [0, 0, -1, 4, 0, 0, 0, 4, 0, 0, 0, -1, 0, 0, 0...             0.312500   
2   [0, 0, 0, -1, 0, -1, 0, 1, 0, 0, 0, 0, 0, 0, 0...            -0.062500   
3    [2, 1, 1, 5, 0, 1, 1, 3, 1, 3, 0, 0, 2, 1, 1, 1]             1.437500   
4   [0, -1, 0, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0, -1, 0...             0.062500   
5   [0, 0, 0, 2, 0, 0, -1, 2, 0, 1, 0, 0, 1, 0, 0, 0]             0.312500   
6   [0, 0, 0, 0, 1, 0, 0, 2, 0, 0, 0, 0, 0, 0, -1, 0]             0.125000   
7   [0, 0, -1, 0, 0, 0, 0, 3, 0, 0, 0, 0, -1, 0, 0...             0.062500   
8   [0, 0, 0, 0, 0, 0, 0, 1, -1, 0, 0, -1, 0, 0, 0...            -0.062500   
9   [1, 0, 0, 3, 0, 0, -1, 2, 1, 0, 0, 0, 0, 0, 0, 0]             0.375000   
10   [0, 0, 0, 3, 1, 0, 0, 1, 1, 0, 0, 1, 0, 0, 0, 1]             0.500000   
11  [0, 0, 0, 1, 0, -1, 0, 2, 0, 0, -1, 0, 0, 0, 0...            -0.062500   
12  [0, 0, -1, 0, 0, 0, -1, 0, -1, 0, 0, 0, 0, 0, ...            -0.187500   
13                                 [0, 1, 2, 0, 0, 0]             0.500000   
14                                 [1, 1, 2, 1, 0, 1]             1.000000   
15                                 [2, 0, 1, 2, 1, 1]             1.166667   
16                                 [2, 0, 5, 0, 1, 1]             1.500000   
17                                 [0, 0, 1, 0, 0, 1]             0.333333   
18                                 [1, 0, 2, 0, 0, 1]             0.666667   
19                                 [3, 0, 2, 2, 2, 4]             2.166667   

     Age     Dataset  Updated Age  
0   {67}  {2016 UGA}           67  
1   {68}  {2017 UGA}           68  
2   {83}  {2017 UGA}           83  
3   {66}  {2017 UGA}           66  
4   {71}  {2017 UGA}           71  
5   {67}  {2017 UGA}           67  
6   {67}  {2017 UGA}           67  
7   {67}  {2017 UGA}           67  
8   {67}  {2017 UGA}           67  
9   {74}  {2017 UGA}           74  
10  {83}  {2017 UGA}           83  
11  {71}  {2017 UGA}           71  
12  {73}  {2017 UGA}           73  
13  {76}  {2019 UGA}           76  
14  {71}  {2019 UGA}           71  
15  {70}  {2019 UGA}           70  
16  {67}  {2019 UGA}           67  
17  {70}  {2019

In [102]:
avg_improvement = sum(group6_one_yr['Average Improvement'])/len(group6_one_yr)
avg_improvement

0.5572916666666667

In [104]:
improvement_ct = 0
for i in range(len(group6_one_yr)):
   if group6_one_yr['Average Improvement'][i] > 0:
       improvement_ct += 1
improvement_ct/len(group6_one_yr)

0.8

In [106]:
one_yr_group_avg = ['N/A', 1.375, 1.0338319088319088, 0.9637045231071779, 0.8615196078431373, 0.5572916666666667]
one_yr_group_improvement = ['N/A', 0.875*100, 0.8717948717948718*100, 0.8584070796460177*100, 0.7647058823529411*100, 0.8*100]

The average improvement in seniors 0.55729. The trends show that as the age of the one year subjects increase, the overall improvement decreases.

In [109]:
group1_multi_yr = [] # Infants(<1 year old)
group2_multi_yr = [] # Children(1-11 years old)
group3_multi_yr = [] # Adolescents(12-17 years old)
group4_multi_yr = [] # Adults(18-44 years old)
group5_multi_yr = [] # Middle Aged Adults(45-64 years old)
group6_multi_yr = [] # Seniors(>65 years old)
for i in range(len(multi_yr_group)):
    if max(multi_yr_group['Age'][i]) < 1:
        group1_multi_yr.append(multi_yr_group.loc[i])
    elif max(multi_yr_group['Age'][i]) < 12:
        group2_multi_yr.append(multi_yr_group.loc[i])
    elif max(multi_yr_group['Age'][i]) < 18:
        group3_multi_yr.append(multi_yr_group.loc[i])
    elif max(multi_yr_group['Age'][i]) < 45:
        group4_multi_yr.append(multi_yr_group.loc[i])
    elif max(multi_yr_group['Age'][i]) < 65:
        group5_multi_yr.append(multi_yr_group.loc[i])
    else:
        group6_multi_yr.append(multi_yr_group.loc[i])

In [111]:
group1_multi_yr

[]

There is no available data for infants who were vaccinated over multiple years.

In [114]:
group2_multi_yr

[]

There is no available data for children who were vaccinated over multiple years.

In [117]:
group3_multi_yr = pd.DataFrame(group3_multi_yr).reset_index()
improvement_arr = []
for i in range(len(group3_multi_yr['Virus_Year_Measurements'])):
    temp_improvement_arr = []
    for virus in group3_multi_yr['Virus_Year_Measurements'][i]:
        for dataset in group3_multi_yr['Virus_Year_Measurements'][i][virus]:
            delta = group3_multi_yr['Virus_Year_Measurements'][i][virus][dataset]
            temp_improvement_arr.append(delta[len(delta)-1] - delta[0])
    improvement_arr.append(temp_improvement_arr)
group3_multi_yr['Improvement'] = improvement_arr
average_arr = []
for i in range(len(group3_multi_yr['Improvement'])):
   average_arr.append(sum(group3_multi_yr['Improvement'][i])/len(group3_multi_yr['Improvement'][i]))
group3_multi_yr['Average Improvement'] = average_arr
group3_multi_yr

level_0  index base_id  \
0          1      1    001C   
1          2      5    004C   
2          4      7    005C   
3          5      9    006C   
4          7     11    007C   
..       ...    ...     ...   
119      246    362    202C   
120      249    365    204C   
121      252    372   207CC   
122      254    374    208C   
123      256    378    210C   

                               Virus_Year_Measurements  \
0    {'H1N1 A/Beijing/262/1995': {'2017 UGA': [0, 2...   
1    {'H1N1 A/Beijing/262/1995': {'2017 UGA': [3, 3...   
2    {'H1N1 A/Beijing/262/1995': {'2017 UGA': [2, 2...   
3    {'H1N1 A/Beijing/262/1995': {'2017 UGA': [3, 4...   
4    {'H1N1 A/Beijing/262/1995': {'2017 UGA': [0, 0...   
..                                                 ...   
119  {'H1N1 A/Brisbane/2/2018': {'2019 UGA': [4, 5,...   
120  {'H1N1 A/Brisbane/2/2018': {'2019 UGA': [3, 5,...   
121  {'H1N1 A/Brisbane/2/2018': {'2019 UGA': [6, 6,...   
122  {'H1N1 A/Brisbane/2/2018': {'2019 UGA': [6, 6]...   
123  {'H1N1 A/Brisbane/2/2018': {'2019 UGA': [5, 5,...   

                                           Improvement  Average Improvement  \
0    [0, 1, 0, 0, 0, -1, 2, -1, 0, 0, 1, 1, 2, -1, ...             0.360000   
1    [0, 0, 0, 1, 1, 0, -1, 0, -1, -1, 1, 0, -4, 0,...             0.055556   
2    [0, 0, 0, 1, 0, 2, 0, -3, 0, 0, 0, 4, -1, 1, 0...             0.440000   
3    [0, 0, 0, 1, 1, -1, 2, 2, -1, 1, -4, 0, 0, 1, ...             0.580645   
4    [0, 0, 0, 0, 0, 0, 0, -1, 0, 0, 0, 0, 0, 0, 0,...             0.032258   
..                                                 ...                  ...   
119    [0, 1, 0, 0, 0, 0, 2, 0, 0, 2, -1, -1, 1, 0, 0]             0.266667   
120    [0, 1, 3, 2, 1, -1, 5, 0, 1, 1, 3, -1, 1, 0, 4]             1.333333   
121                [0, 0, 0, 0, 0, 0, -1, 0, -1, 1, 1]             0.000000   
122                     [0, 4, 0, 0, 2, 0, 1, 3, 0, 5]             1.500000   
123     [1, 1, 0, 1, 0, -3, 4, 0, 2, 2, 0, 1, 2, 1, 1]             0.866667   

                  Age                                   Dataset  
0            {16, 15}                      {2017 UGA, 2018 UGA}  
1    {16, 17, 14, 15}  {2017 UGA, 2019 UGA, 2018 UGA, 2020 UGA}  
2            {16, 17}                      {2017 UGA, 2018 UGA}  
3        {16, 17, 15}            {2017 UGA, 2019 UGA, 2018 UGA}  
4        {16, 17, 15}            {2017 UGA, 2019 UGA, 2018 UGA}  
..                ...                                       ...  
119          {13, 14}            {2019 UGA, 2020 UGA, 2021 UGA}  
120          {12, 13}            {2019 UGA, 2020 UGA, 2021 UGA}  
121          {12, 13}                      {2019 UGA, 2020 UGA}  
122          {12, 14}                      {2019 UGA, 2021 UGA}  
123      {11, 12, 13}            {2019 UGA, 2020 UGA, 2021 UGA}  

[124 rows x 8 columns]

In [119]:
avg_improvement = sum(group3_multi_yr['Average Improvement'])/len(group3_multi_yr)
avg_improvement

0.2741436485969319

In [121]:
improvement_ct = 0
for i in range(len(group3_multi_yr)):
   if group3_multi_yr['Average Improvement'][i] > 0:
       improvement_ct += 1
improvement_ct/len(group3_multi_yr)

0.7741935483870968

The average improvement in adolescents who were vaccinated over multiple years is 0.2741.

In [124]:
group4_multi_yr = pd.DataFrame(group4_multi_yr).reset_index()
improvement_arr = []
for i in range(len(group4_multi_yr['Virus_Year_Measurements'])):
    temp_improvement_arr = []
    for virus in group4_multi_yr['Virus_Year_Measurements'][i]:
        for dataset in group4_multi_yr['Virus_Year_Measurements'][i][virus]:
            delta = group4_multi_yr['Virus_Year_Measurements'][i][virus][dataset]
            temp_improvement_arr.append(delta[len(delta)-1] - delta[0])
    improvement_arr.append(temp_improvement_arr)
group4_multi_yr['Improvement'] = improvement_arr
average_arr = []
for i in range(len(group4_multi_yr['Improvement'])):
   average_arr.append(sum(group4_multi_yr['Improvement'][i])/len(group4_multi_yr['Improvement'][i]))
group4_multi_yr['Average Improvement'] = average_arr
group4_multi_yr

level_0  index base_id  \
0          0      0     001   
1          3      6     005   
2          6     10     007   
3         11     16     010   
4         12     18     011   
..       ...    ...     ...   
146      476    719     524   
147      478    721     526   
148      479    726     531   
149      481    728     539   
150      482    729     540   

                               Virus_Year_Measurements  \
0    {'H1N1 A/Beijing/262/1995': {'2016 UGA': [3, 3...   
1    {'H1N1 A/Beijing/262/1995': {'2016 UGA': [1, 0...   
2    {'H1N1 A/Beijing/262/1995': {'2016 UGA': [4, 4...   
3    {'H1N1 A/Beijing/262/1995': {'2016 UGA': [1, 2...   
4    {'H1N1 A/Beijing/262/1995': {'2016 UGA': [1, 2...   
..                                                 ...   
146  {'H1N1 A/Brisbane/2/2018': {'2019 UGA': [5, 4,...   
147  {'H1N1 A/Brisbane/2/2018': {'2019 UGA': [5, 4,...   
148  {'H1N1 A/Brisbane/2/2018': {'2019 UGA': [2, 2,...   
149  {'H1N1 A/Brisbane/2/2018': {'2020 UGA': [0, 5,...   
150  {'H1N1 A/Brisbane/2/2018': {'2020 UGA': [3, 5,...   

                                           Improvement  Average Improvement  \
0    [0, 0, 0, 0, 4, 2, 5, 1, 0, 0, 4, 1, 1, 0, 1, ...             1.161290   
1    [-1, 0, 0, 0, 0, -1, 0, 2, 1, -1, 0, 0, 0, 0, ...             0.000000   
2    [-4, 1, 1, 0, 0, 0, 0, 0, 4, -1, 0, 0, 0, 0, 0...             0.170732   
3    [-1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,...             0.156250   
4    [-1, 0, 0, 0, 0, -1, 2, 1, 0, -1, 0, 0, 0, -1,...             0.125000   
..                                                 ...                  ...   
146      [1, 1, 1, 0, 1, 0, 1, 0, 1, 2, 0, 1, 1, 3, 0]             0.866667   
147    [1, 1, 0, 0, 0, -2, 1, 0, 0, 0, 0, 0, 0, -1, 0]             0.000000   
148      [2, 0, 3, 1, 3, 0, 3, 0, 2, 2, 1, 1, 3, 0, 3]             1.600000   
149                        [5, 0, 1, 0, 4, 0, 2, 1, 0]             1.444444   
150                        [1, 2, 1, 3, 2, 2, 3, 0, 3]             1.888889   

                          Age  \
0            {32, 33, 29, 31}   
1        {25, 26, 27, 28, 29}   
2                    {22, 23}   
3                    {25, 26}   
4    {25, 26, 27, 28, 29, 30}   
..                        ...   
146              {24, 25, 23}   
147                  {27, 28}   
148              {19, 20, 21}   
149                      {18}   
150                      {18}   

                                               Dataset  
0             {2016 UGA, 2019 UGA, 2021 UGA, 2020 UGA}  
1    {2021 UGA, 2016 UGA, 2020 UGA, 2018 UGA, 2017 ...  
2                       {2017 UGA, 2016 UGA, 2018 UGA}  
3                                 {2017 UGA, 2016 UGA}  
4    {2021 UGA, 2016 UGA, 2020 UGA, 2018 UGA, 2017 ...  
..                                                 ...  
146                     {2019 UGA, 2020 UGA, 2021 UGA}  
147                     {2019 UGA, 2020 UGA, 2021 UGA}  
148                     {2019 UGA, 2020 UGA, 2021 UGA}  
149                               {2020 UGA, 2021 UGA}  
150                               {2020 UGA, 2021 UGA}  

[151 rows x 8 columns]

In [126]:
avg_improvement = sum(group4_multi_yr['Average Improvement'])/len(group4_multi_yr)
avg_improvement

0.4064528312752101

In [128]:
improvement_ct = 0
for i in range(len(group4_multi_yr)):
   if group4_multi_yr['Average Improvement'][i] > 0:
       improvement_ct += 1
improvement_ct/len(group4_multi_yr)

0.7682119205298014

The average improvement in adults 18-44 who were vaccinated over multuple years is 0.40645.

In [131]:
group5_multi_yr = pd.DataFrame(group5_multi_yr).reset_index()
improvement_arr = []
for i in range(len(group5_multi_yr['Virus_Year_Measurements'])):
    temp_improvement_arr = []
    for virus in group5_multi_yr['Virus_Year_Measurements'][i]:
        for dataset in group5_multi_yr['Virus_Year_Measurements'][i][virus]:
            delta = group5_multi_yr['Virus_Year_Measurements'][i][virus][dataset]
            temp_improvement_arr.append(delta[len(delta)-1] - delta[0])
    improvement_arr.append(temp_improvement_arr)
group5_multi_yr['Improvement'] = improvement_arr
average_arr = []
for i in range(len(group5_multi_yr['Improvement'])):
   average_arr.append(sum(group5_multi_yr['Improvement'][i])/len(group5_multi_yr['Improvement'][i]))
group5_multi_yr['Average Improvement'] = average_arr
group5_multi_yr

level_0  index base_id  \
0          8     12     008   
1         21     30     017   
2         49     67     037   
3         51     69     038   
4         70     93     051   
..       ...    ...     ...   
105      456    677     480   
106      459    682     485   
107      469    699     502   
108      473    714     518   
109      474    715     519   

                               Virus_Year_Measurements  \
0    {'H1N1 A/Beijing/262/1995': {'2016 UGA': [2, 2...   
1    {'H1N1 A/Beijing/262/1995': {'2016 UGA': [3, 2...   
2    {'H1N1 A/Beijing/262/1995': {'2016 UGA': [2, 2...   
3    {'H1N1 A/Beijing/262/1995': {'2016 UGA': [2, 3...   
4    {'H1N1 A/Beijing/262/1995': {'2016 UGA': [1, 1...   
..                                                 ...   
105  {'H1N1 A/Brisbane/2/2018': {'2019 UGA': [0, 3,...   
106  {'H1N1 A/Brisbane/2/2018': {'2019 UGA': [3, 3,...   
107  {'H1N1 A/Brisbane/2/2018': {'2019 UGA': [0, 4,...   
108  {'H1N1 A/Brisbane/2/2018': {'2019 UGA': [1, 2,...   
109  {'H1N1 A/Brisbane/2/2018': {'2019 UGA': [4, 5]...   

                                           Improvement  Average Improvement  \
0    [0, -1, 0, 0, 0, -1, 1, 1, -1, -1, 0, 3, 1, -1...            -0.035714   
1    [-2, 1, 0, 0, 1, 0, 0, 1, -1, 0, -1, 1, 0, -1,...            -0.125000   
2    [-1, 0, 0, 6, 0, 1, 0, -1, 0, 0, 4, 0, -2, 2, ...             0.446809   
3    [1, -1, 0, -2, 0, -1, -1, 0, 1, 0, -2, 0, 4, 0...             0.392857   
4    [0, 0, 0, 4, 0, 0, 0, 1, 0, 0, 0, 0, 6, -1, -5...             0.357143   
..                                                 ...                  ...   
105     [0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, -2, 2]             0.133333   
106              [-1, 1, 0, 3, 1, 0, -1, 1, -1, 0, -1]             0.181818   
107  [3, 0, -3, 0, 3, -1, 2, 0, 2, -1, 1, 3, 0, -7, 3]             0.333333   
108                  [4, 1, 1, 5, 0, 0, 3, 0, 3, 0, 1]             1.636364   
109                     [1, 1, 2, 4, 1, 0, 2, 1, 3, 1]             1.600000   

                          Age  \
0    {46, 47, 48, 49, 50, 51}   
1    {51, 52, 53, 54, 55, 56}   
2            {57, 58, 59, 60}   
3            {43, 44, 46, 47}   
4    {45, 46, 47, 48, 49, 50}   
..                        ...   
105              {51, 52, 53}   
106                  {58, 59}   
107                  {53, 54}   
108                  {49, 50}   
109                  {49, 47}   

                                               Dataset  
0    {2021 UGA, 2016 UGA, 2020 UGA, 2018 UGA, 2017 ...  
1    {2021 UGA, 2016 UGA, 2020 UGA, 2018 UGA, 2017 ...  
2             {2017 UGA, 2016 UGA, 2018 UGA, 2019 UGA}  
3    {2021 UGA, 2016 UGA, 2020 UGA, 2018 UGA, 2017 ...  
4    {2021 UGA, 2016 UGA, 2020 UGA, 2018 UGA, 2017 ...  
..                                                 ...  
105                     {2019 UGA, 2020 UGA, 2021 UGA}  
106                               {2019 UGA, 2020 UGA}  
107                     {2019 UGA, 2020 UGA, 2021 UGA}  
108                               {2019 UGA, 2020 UGA}  
109                               {2019 UGA, 2021 UGA}  

[110 rows x 8 columns]

In [133]:
avg_improvement = sum(group5_multi_yr['Average Improvement'])/len(group5_multi_yr)
avg_improvement

0.569363267456103

In [135]:
improvement_ct = 0
for i in range(len(group5_multi_yr)):
   if group5_multi_yr['Average Improvement'][i] > 0:
       improvement_ct += 1
improvement_ct/len(group5_multi_yr)

0.8454545454545455

The average improvement in adults aged 45-62 who were vaccinated over multiple years is 0.569.

In [138]:
group6_multi_yr = pd.DataFrame(group6_multi_yr).reset_index()
improvement_arr = []
for i in range(len(group6_multi_yr['Virus_Year_Measurements'])):
    temp_improvement_arr = []
    for virus in group6_multi_yr['Virus_Year_Measurements'][i]:
        for dataset in group6_multi_yr['Virus_Year_Measurements'][i][virus]:
            delta = group6_multi_yr['Virus_Year_Measurements'][i][virus][dataset]
            temp_improvement_arr.append(delta[len(delta)-1] - delta[0])
    improvement_arr.append(temp_improvement_arr)
group6_multi_yr['Improvement'] = improvement_arr
average_arr = []
for i in range(len(group6_multi_yr['Improvement'])):
   average_arr.append(sum(group6_multi_yr['Improvement'][i])/len(group6_multi_yr['Improvement'][i]))
group6_multi_yr['Average Improvement'] = average_arr
group6_multi_yr

level_0  index base_id                            Virus_Year_Measurements  \
0        15     24     014  {'H1N1 A/Beijing/262/1995': {'2016 UGA': [1, 1...   
1        63     86     047  {'H1N1 A/Beijing/262/1995': {'2016 UGA': [0, 0...   
2       106    143     077  {'H1N1 A/Beijing/262/1995': {'2016 UGA': [1, 0...   
3       110    154     083  {'H1N1 A/Beijing/262/1995': {'2016 UGA': [0, 0...   
4       121    170     091  {'H1N1 A/Beijing/262/1995': {'2016 UGA': [0, 0...   
..      ...    ...     ...                                                ...   
93      453    673     476  {'H1N1 A/Brisbane/2/2018': {'2019 UGA': [0, 0,...   
94      457    678     481  {'H1N1 A/Brisbane/2/2018': {'2019 UGA': [0, 5]...   
95      458    681     484  {'H1N1 A/Brisbane/2/2018': {'2019 UGA': [2, 1,...   
96      477    720     525  {'H1N1 A/Brisbane/2/2018': {'2019 UGA': [0, 3,...   
97      480    727     535  {'H1N1 A/Brisbane/2/2018': {'2019 UGA': [1, 6]...   

                                          Improvement  Average Improvement  \
0   [-1, 0, 0, 0, 0, 0, 1, -1, 0, 0, 1, 0, 1, 0, 0...             0.000000   
1   [0, 0, 0, 0, 0, 1, 0, -1, 0, 0, 0, 0, 0, 1, -2...             0.017857   
2   [-1, 0, 0, 0, 0, 0, 0, 1, 1, -1, 0, 0, 0, -1, ...            -0.071429   
3   [0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, -1, 0,...            -0.017857   
4   [0, 0, 0, 0, 1, 0, 0, 0, 0, -1, 0, -1, -1, 0, ...            -0.048780   
..                                                ...                  ...   
93    [2, 0, 0, 0, 0, -1, 1, 0, 0, 0, 0, 0, 1, -2, 0]             0.066667   
94                     [5, 4, 1, 5, 4, 2, 4, 3, 6, 3]             3.700000   
95              [-1, 1, 0, -1, 1, 0, -1, 1, -2, 1, 2]             0.090909   
96                  [4, 3, 0, 3, 4, 0, 1, 5, 5, 3, 2]             2.727273   
97                   [5, 0, 0, 1, -1, 0, 5, -1, 4, 0]             1.300000   

                         Age  \
0           {64, 65, 66, 62}   
1   {64, 65, 66, 67, 62, 63}   
2   {64, 65, 66, 67, 68, 69}   
3       {66, 67, 68, 69, 70}   
4               {72, 73, 74}   
..                       ...   
93              {67, 68, 69}   
94                  {67, 69}   
95                  {70, 71}   
96                  {64, 65}   
97                  {65, 63}   

                                              Dataset  
0   {2021 UGA, 2016 UGA, 2020 UGA, 2017 UGA, 2019 ...  
1   {2021 UGA, 2016 UGA, 2020 UGA, 2018 UGA, 2017 ...  
2   {2021 UGA, 2016 UGA, 2020 UGA, 2018 UGA, 2017 ...  
3   {2021 UGA, 2016 UGA, 2020 UGA, 2018 UGA, 2017 ...  
4                      {2017 UGA, 2016 UGA, 2018 UGA}  
..                                                ...  
93                     {2019 UGA, 2020 UGA, 2021 UGA}  
94                               {2019 UGA, 2021 UGA}  
95                               {2019 UGA, 2020 UGA}  
96                               {2019 UGA, 2020 UGA}  
97                               {2019 UGA, 2021 UGA}  

[98 rows x 8 columns]

In [140]:
avg_improvement = sum(group6_multi_yr['Average Improvement'])/len(group6_multi_yr)
avg_improvement

0.32926395845086126

In [142]:
improvement_ct = 0
for i in range(len(group6_multi_yr)):
   if group6_multi_yr['Average Improvement'][i] > 0:
       improvement_ct += 1
improvement_ct/len(group6_multi_yr)

0.6938775510204082

In [144]:
multi_yr_group_avg = ['N/A', 'N/A', 0.2741436485969319, 0.4064528312752101, 0.569363267456103, 0.32926395845086126]
multi_yr_group_improvement = ['N/A', 'N/A', 0.7741935483870968*100, 0.7682119205298014*100, 0.8454545454545455*100, 0.6938775510204082*100]

In [146]:
groups = pd.DataFrame(one_yr_group_avg, columns = ['One-Year Group Average Improvement'])
groups['One-Year Group Improvement Percentage(%)'] = one_yr_group_improvement
groups['Multi-Year Group Average Improvement'] = multi_yr_group_avg
groups['Multi-Year Group Improvement Percentage(%)'] = multi_yr_group_improvement
groups.index = ['Group 1(Infants (<0))', 'Group 2(Children(1-11))', 'Group 3(Adolescents(12-17))', 'Group 4(Adults(18-44))', 'Group 5(Middle-Aged Adults(45-62))', 'Group 6(Seniors(>63))']
groups

One-Year Group Average Improvement  \
Group 1(Infants (<0))                                             N/A   
Group 2(Children(1-11))                                         1.375   
Group 3(Adolescents(12-17))                                  1.033832   
Group 4(Adults(18-44))                                       0.963705   
Group 5(Middle-Aged Adults(45-62))                            0.86152   
Group 6(Seniors(>63))                                        0.557292   

                                   One-Year Group Improvement Percentage(%)  \
Group 1(Infants (<0))                                                   N/A   
Group 2(Children(1-11))                                                87.5   
Group 3(Adolescents(12-17))                                       87.179487   
Group 4(Adults(18-44))                                            85.840708   
Group 5(Middle-Aged Adults(45-62))                                76.470588   
Group 6(Seniors(>63))                                                  80.0   

                                   Multi-Year Group Average Improvement  \
Group 1(Infants (<0))                                               N/A   
Group 2(Children(1-11))                                             N/A   
Group 3(Adolescents(12-17))                                    0.274144   
Group 4(Adults(18-44))                                         0.406453   
Group 5(Middle-Aged Adults(45-62))                             0.569363   
Group 6(Seniors(>63))                                          0.329264   

                                   Multi-Year Group Improvement Percentage(%)  
Group 1(Infants (<0))                                                     N/A  
Group 2(Children(1-11))                                                   N/A  
Group 3(Adolescents(12-17))                                         77.419355  
Group 4(Adults(18-44))                                              76.821192  
Group 5(Middle-Aged Adults(45-62))                                  84.545455  
Group 6(Seniors(>63))                                               69.387755

The average improvement in seniors is 0.326. The adolescents have the lowest improvement and seniors have second lowest. The older adults ahve higher improvement than the younger ones. Now the cross reactivity of each vaccine must be studied.

In [171]:
def create_virus_dict_cross(group):
    virus_dict = {}
    for virus, virus_group in group.groupby('Virus'):
        virus_dict[virus] = virus_group.sort_values('Time')['Measurement'].tolist()
    return virus_dict

In [173]:
study_2016 = []
study_2017 = []
study_2018 = []
study_2019 = []
study_2020 = []
study_2021 = []

for i in range(len(df)):
    if df['Dataset'][i] == '2016 UGA':
        study_2016.append(df.loc[i])
    elif df['Dataset'][i] == '2017 UGA':
        study_2017.append(df.loc[i])
    elif df['Dataset'][i] == '2018 UGA':
        study_2018.append(df.loc[i])
    elif df['Dataset'][i] == '2019 UGA':
        study_2019.append(df.loc[i])
    elif df['Dataset'][i] == '2020 UGA':
        study_2020.append(df.loc[i])
    else:
        study_2021.append(df.loc[i])

The 2016 study will be considered first.

In [175]:
study_2016 = pd.DataFrame(study_2016)
study_2016

Dataset                          Virus         Subject  Age    Time  \
0     2016 UGA   H1N1 A/South Carolina/1/1918  ID_001_UGA2016   29    Day0   
1     2016 UGA          H1N1 A/Weiss/JY2/1943  ID_001_UGA2016   29    Day0   
2     2016 UGA    H1N1 A/Fort Monmouth/1/1947  ID_001_UGA2016   29    Day0   
3     2016 UGA           H1N1 A/Denver/1/1957  ID_001_UGA2016   29    Day0   
4     2016 UGA       H1N1 A/New Jersey/8/1976  ID_001_UGA2016   29    Day0   
...        ...                            ...             ...  ...     ...   
6363  2016 UGA   H1N1 A/New Caledonia/20/1999  ID_151_UGA2016   66  Day250   
6364  2016 UGA  H1N1 A/Solomon Islands/3/2006  ID_151_UGA2016   66  Day250   
6365  2016 UGA        H1N1 A/Brisbane/59/2007  ID_151_UGA2016   66  Day250   
6366  2016 UGA       H1N1 A/California/7/2009  ID_151_UGA2016   66  Day250   
6367  2016 UGA        H1N1 A/Michigan/45/2015  ID_151_UGA2016   66  Day250   

      Measurement base_id  
0               2     001  
1               4     001  
2               1     001  
3               0     001  
4               0     001  
...           ...     ...  
6363            1     151  
6364            1     151  
6365            1     151  
6366            3     151  
6367            2     151  

[6368 rows x 7 columns]

In [176]:
result_2016 = study_2016.groupby('base_id').apply(create_virus_dict_cross).reset_index(name='Virus Measurements')
result_2016

C:\Users\anall\AppData\Local\Temp\ipykernel_35012\4087756756.py:1: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  result_2016 = study_2016.groupby('base_id').apply(create_virus_dict_cross).reset_index(name='Virus Measurements')


base_id                                 Virus Measurements
0       001  {'H1N1 A/Beijing/262/1995': [3, 3], 'H1N1 A/Br...
1       002  {'H1N1 A/Beijing/262/1995': [4, 3], 'H1N1 A/Br...
2       003  {'H1N1 A/Beijing/262/1995': [4, 3], 'H1N1 A/Br...
3       004  {'H1N1 A/Beijing/262/1995': [1, 3], 'H1N1 A/Br...
4       005  {'H1N1 A/Beijing/262/1995': [1, 0, 0], 'H1N1 A...
..      ...                                                ...
143     147  {'H1N1 A/Beijing/262/1995': [3, 3, 1], 'H1N1 A...
144     148  {'H1N1 A/Beijing/262/1995': [1, 0, 0], 'H1N1 A...
145     149  {'H1N1 A/Beijing/262/1995': [1, 0, 0], 'H1N1 A...
146     150  {'H1N1 A/Beijing/262/1995': [0, 0, 0], 'H1N1 A...
147     151  {'H1N1 A/Beijing/262/1995': [2, 0, 0], 'H1N1 A...

[148 rows x 2 columns]

In [177]:
improvement_arr = []
for i in range(len(result_2016['Virus Measurements'])):
    temp_improvement_arr = []
    for virus in result_2016['Virus Measurements'][i]:
        delta = result_2016['Virus Measurements'][i][virus]
        temp_improvement_arr.append(delta[len(delta)-1]-delta[0])
    improvement_arr.append(temp_improvement_arr)

In [178]:
result_2016['Improvements'] = improvement_arr
result_2016

base_id                                 Virus Measurements  \
0       001  {'H1N1 A/Beijing/262/1995': [3, 3], 'H1N1 A/Br...   
1       002  {'H1N1 A/Beijing/262/1995': [4, 3], 'H1N1 A/Br...   
2       003  {'H1N1 A/Beijing/262/1995': [4, 3], 'H1N1 A/Br...   
3       004  {'H1N1 A/Beijing/262/1995': [1, 3], 'H1N1 A/Br...   
4       005  {'H1N1 A/Beijing/262/1995': [1, 0, 0], 'H1N1 A...   
..      ...                                                ...   
143     147  {'H1N1 A/Beijing/262/1995': [3, 3, 1], 'H1N1 A...   
144     148  {'H1N1 A/Beijing/262/1995': [1, 0, 0], 'H1N1 A...   
145     149  {'H1N1 A/Beijing/262/1995': [1, 0, 0], 'H1N1 A...   
146     150  {'H1N1 A/Beijing/262/1995': [0, 0, 0], 'H1N1 A...   
147     151  {'H1N1 A/Beijing/262/1995': [2, 0, 0], 'H1N1 A...   

                                          Improvements  
0     [0, 0, 2, 1, 1, 0, 1, 0, 1, 3, 0, 1, 1, 0, 2, 0]  
1    [-1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, -1, 0, 0, 0...  
2    [-1, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 3, 1, 1]  
3     [2, 0, 2, 4, 0, 1, 0, 4, 2, 2, 2, 2, 3, 0, 0, 4]  
4    [-1, 0, 1, 0, 0, 0, 0, 1, 1, 0, -1, 1, 0, 2, 0...  
..                                                 ...  
143  [-2, 1, 0, 1, 0, 1, 0, 2, 0, 0, 0, -1, 1, 3, 1...  
144  [-1, 1, 0, 0, 0, 1, 0, 2, -2, 0, -5, -4, 0, 0,...  
145  [-1, 0, 0, -1, 0, 0, 0, 0, -1, 0, -4, -2, 0, -...  
146  [0, 0, 0, 1, 0, 0, 0, 0, 0, 4, -2, 0, 0, 0, 0, 0]  
147  [-2, 1, -1, 1, 1, 0, 0, 1, -1, 0, 0, -3, 0, 0,...  

[148 rows x 3 columns]

In [179]:
avg_improvement = []
for i in range(len(result_2016['Improvements'])):
    avg_improvement.append(sum(result_2016['Improvements'][i]) / len(result_2016['Improvements'][i]))
result_2016['Average Improvement'] = avg_improvement

In [180]:
result_2016

base_id                                 Virus Measurements  \
0       001  {'H1N1 A/Beijing/262/1995': [3, 3], 'H1N1 A/Br...   
1       002  {'H1N1 A/Beijing/262/1995': [4, 3], 'H1N1 A/Br...   
2       003  {'H1N1 A/Beijing/262/1995': [4, 3], 'H1N1 A/Br...   
3       004  {'H1N1 A/Beijing/262/1995': [1, 3], 'H1N1 A/Br...   
4       005  {'H1N1 A/Beijing/262/1995': [1, 0, 0], 'H1N1 A...   
..      ...                                                ...   
143     147  {'H1N1 A/Beijing/262/1995': [3, 3, 1], 'H1N1 A...   
144     148  {'H1N1 A/Beijing/262/1995': [1, 0, 0], 'H1N1 A...   
145     149  {'H1N1 A/Beijing/262/1995': [1, 0, 0], 'H1N1 A...   
146     150  {'H1N1 A/Beijing/262/1995': [0, 0, 0], 'H1N1 A...   
147     151  {'H1N1 A/Beijing/262/1995': [2, 0, 0], 'H1N1 A...   

                                          Improvements  Average Improvement  
0     [0, 0, 2, 1, 1, 0, 1, 0, 1, 3, 0, 1, 1, 0, 2, 0]               0.8125  
1    [-1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, -1, 0, 0, 0...              -0.0625  
2    [-1, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 3, 1, 1]               0.3750  
3     [2, 0, 2, 4, 0, 1, 0, 4, 2, 2, 2, 2, 3, 0, 0, 4]               1.7500  
4    [-1, 0, 1, 0, 0, 0, 0, 1, 1, 0, -1, 1, 0, 2, 0...               0.3125  
..                                                 ...                  ...  
143  [-2, 1, 0, 1, 0, 1, 0, 2, 0, 0, 0, -1, 1, 3, 1...               0.5000  
144  [-1, 1, 0, 0, 0, 1, 0, 2, -2, 0, -5, -4, 0, 0,...              -0.4375  
145  [-1, 0, 0, -1, 0, 0, 0, 0, -1, 0, -4, -2, 0, -...              -0.8750  
146  [0, 0, 0, 1, 0, 0, 0, 0, 0, 4, -2, 0, 0, 0, 0, 0]               0.1875  
147  [-2, 1, -1, 1, 1, 0, 0, 1, -1, 0, 0, -3, 0, 0,...              -0.1875  

[148 rows x 4 columns]

In [181]:
sum(avg_improvement) / len(avg_improvement)

0.2951858108108108

In [229]:
improvement_ct = 0
for i in range(len(result_2016)):
   if result_2016['Average Improvement'][i] > 0:
       improvement_ct += 1
improvement_ct/len(result_2016)

0.6351351351351351

The average improvement of patients in 2016 was 0.29518

In [183]:
study_2017 = pd.DataFrame(study_2017).reset_index()
result_2017 = study_2017.groupby('base_id').apply(create_virus_dict_cross).reset_index(name='Virus Measurements')

C:\Users\anall\AppData\Local\Temp\ipykernel_35012\1449408175.py:2: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  result_2017 = study_2017.groupby('base_id').apply(create_virus_dict_cross).reset_index(name='Virus Measurements')


In [185]:
improvement_arr = []
for i in range(len(result_2017['Virus Measurements'])):
    temp_improvement_arr = []
    for virus in result_2017['Virus Measurements'][i]:
        delta = result_2017['Virus Measurements'][i][virus]
        temp_improvement_arr.append(delta[len(delta)-1]-delta[0])
    improvement_arr.append(temp_improvement_arr)

In [187]:
result_2017['Improvement'] = improvement_arr
result_2017

base_id                                 Virus Measurements  \
0      001C  {'H1N1 A/Beijing/262/1995': [0, 2, 0], 'H1N1 A...   
1      004C  {'H1N1 A/Beijing/262/1995': [3, 3, 3], 'H1N1 A...   
2       005  {'H1N1 A/Beijing/262/1995': [0, 0, 0], 'H1N1 A...   
3      005C  {'H1N1 A/Beijing/262/1995': [2, 2, 2], 'H1N1 A...   
4      006C  {'H1N1 A/Beijing/262/1995': [3, 4, 3], 'H1N1 A...   
..      ...                                                ...   
266     256  {'H1N1 A/Beijing/262/1995': [0, 0], 'H1N1 A/Br...   
267     257  {'H1N1 A/Beijing/262/1995': [1, 0], 'H1N1 A/Br...   
268     260  {'H1N1 A/Beijing/262/1995': [1, 1], 'H1N1 A/Br...   
269     261  {'H1N1 A/Beijing/262/1995': [1, 0], 'H1N1 A/Br...   
270     262  {'H1N1 A/Beijing/262/1995': [4, 5], 'H1N1 A/Br...   

                                           Improvement  
0    [0, 0, 0, -1, -1, 0, 1, 1, -1, 3, 1, -1, 2, 0,...  
1    [0, 0, 0, -1, -4, 0, 1, 1, 0, 1, -1, 0, 1, -1,...  
2    [0, 0, -1, 0, 0, 0, 0, -1, -1, 0, 0, -1, 0, 0,...  
3    [0, 0, 1, 2, -3, 0, 0, 4, 1, 3, 0, 1, 3, 0, 0, 0]  
4    [0, 0, 1, 2, -4, 1, 0, 3, 0, 6, -1, 0, 5, 0, 1...  
..                                                 ...  
266  [0, 1, 0, 3, 0, 0, 1, 4, 0, 3, 1, -1, 2, 1, 0, 1]  
267  [-1, 0, 0, 3, 1, 1, 1, 5, 1, 4, 2, 0, 3, 1, 0, 0]  
268   [0, 1, 3, 3, 0, 3, 0, 4, 3, 2, 4, 3, 1, 4, 1, 0]  
269  [-1, 2, 0, 2, 1, 3, 2, 4, 0, 2, 3, -2, 2, 2, 1...  
270   [1, 1, 1, 0, 1, 1, 1, 0, 1, 1, 1, 0, 1, 1, 0, 1]  

[271 rows x 3 columns]

In [190]:
avg_improvement = []
for i in range(len(result_2017['Improvement'])):
    avg_improvement.append(sum(result_2017['Improvement'][i]) / len(result_2017['Improvement'][i]))
result_2017['Average Improvement'] = avg_improvement
result_2017

base_id                                 Virus Measurements  \
0      001C  {'H1N1 A/Beijing/262/1995': [0, 2, 0], 'H1N1 A...   
1      004C  {'H1N1 A/Beijing/262/1995': [3, 3, 3], 'H1N1 A...   
2       005  {'H1N1 A/Beijing/262/1995': [0, 0, 0], 'H1N1 A...   
3      005C  {'H1N1 A/Beijing/262/1995': [2, 2, 2], 'H1N1 A...   
4      006C  {'H1N1 A/Beijing/262/1995': [3, 4, 3], 'H1N1 A...   
..      ...                                                ...   
266     256  {'H1N1 A/Beijing/262/1995': [0, 0], 'H1N1 A/Br...   
267     257  {'H1N1 A/Beijing/262/1995': [1, 0], 'H1N1 A/Br...   
268     260  {'H1N1 A/Beijing/262/1995': [1, 1], 'H1N1 A/Br...   
269     261  {'H1N1 A/Beijing/262/1995': [1, 0], 'H1N1 A/Br...   
270     262  {'H1N1 A/Beijing/262/1995': [4, 5], 'H1N1 A/Br...   

                                           Improvement  Average Improvement  
0    [0, 0, 0, -1, -1, 0, 1, 1, -1, 3, 1, -1, 2, 0,...               0.3125  
1    [0, 0, 0, -1, -4, 0, 1, 1, 0, 1, -1, 0, 1, -1,...              -0.0625  
2    [0, 0, -1, 0, 0, 0, 0, -1, -1, 0, 0, -1, 0, 0,...              -0.3125  
3    [0, 0, 1, 2, -3, 0, 0, 4, 1, 3, 0, 1, 3, 0, 0, 0]               0.7500  
4    [0, 0, 1, 2, -4, 1, 0, 3, 0, 6, -1, 0, 5, 0, 1...               0.9375  
..                                                 ...                  ...  
266  [0, 1, 0, 3, 0, 0, 1, 4, 0, 3, 1, -1, 2, 1, 0, 1]               1.0000  
267  [-1, 0, 0, 3, 1, 1, 1, 5, 1, 4, 2, 0, 3, 1, 0, 0]               1.3125  
268   [0, 1, 3, 3, 0, 3, 0, 4, 3, 2, 4, 3, 1, 4, 1, 0]               2.0000  
269  [-1, 2, 0, 2, 1, 3, 2, 4, 0, 2, 3, -2, 2, 2, 1...               1.3750  
270   [1, 1, 1, 0, 1, 1, 1, 0, 1, 1, 1, 0, 1, 1, 0, 1]               0.7500  

[271 rows x 4 columns]

In [195]:
sum(avg_improvement)/len(avg_improvement)

0.21309963099630996

In [231]:
improvement_ct = 0
for i in range(len(result_2017)):
   if result_2017['Average Improvement'][i] > 0:
       improvement_ct += 1
improvement_ct/len(result_2017)

0.5018450184501845

The average improvement of subjects in 2017 was 0.213.

In [197]:
study_2018 = pd.DataFrame(study_2018).reset_index()
result_2018 = study_2018.groupby('base_id').apply(create_virus_dict_cross).reset_index(name='Virus Measurements')
improvement_arr = []
for i in range(len(result_2018['Virus Measurements'])):
    temp_improvement_arr = []
    for virus in result_2018['Virus Measurements'][i]:
        delta = result_2018['Virus Measurements'][i][virus]
        temp_improvement_arr.append(delta[len(delta)-1]-delta[0])
    improvement_arr.append(temp_improvement_arr)

result_2018['Improvement'] = improvement_arr
avg_improvement = []
for i in range(len(result_2018['Improvement'])):
    avg_improvement.append(sum(result_2018['Improvement'][i]) / len(result_2018['Improvement'][i]))

result_2018['Average Improvement'] = avg_improvement

C:\Users\anall\AppData\Local\Temp\ipykernel_35012\3149123489.py:2: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  result_2018 = study_2018.groupby('base_id').apply(create_virus_dict_cross).reset_index(name='Virus Measurements')


In [198]:
result_2018

base_id                                 Virus Measurements  \
0      001C  {'H1N1 A/Beijing/262/1995': [0, 1], 'H1N1 A/Br...   
1      004C  {'H1N1 A/Beijing/262/1995': [3, 3], 'H1N1 A/Br...   
2       005  {'H1N1 A/Beijing/262/1995': [0, 0], 'H1N1 A/Br...   
3      005C  {'H1N1 A/Beijing/262/1995': [2, 2], 'H1N1 A/Br...   
4      006C  {'H1N1 A/Beijing/262/1995': [3, 3], 'H1N1 A/Br...   
..      ...                                                ...   
245     252  {'H1N1 A/Beijing/262/1995': [2, 2], 'H1N1 A/Br...   
246     263  {'H1N1 A/Beijing/262/1995': [1, 1], 'H1N1 A/Br...   
247     264  {'H1N1 A/Beijing/262/1995': [2, 3], 'H1N1 A/Br...   
248     265  {'H1N1 A/Beijing/262/1995': [3, 3], 'H1N1 A/Br...   
249     266  {'H1N1 A/Beijing/262/1995': [2, 2], 'H1N1 A/Br...   

                        Improvement  Average Improvement  
0      [1, 0, 2, 0, 2, 0, -1, 0, 0]             0.444444  
1    [0, -1, -1, 0, 0, 0, -1, 0, 0]            -0.333333  
2    [0, 0, -1, 0, -1, 0, 0, 0, -1]            -0.333333  
3      [0, 0, 0, 0, -1, 0, 0, 0, 0]            -0.111111  
4     [0, -1, -1, 0, 0, 0, 0, 0, 0]            -0.222222  
..                              ...                  ...  
245  [0, -1, -1, 0, -1, 0, 0, 0, 0]            -0.333333  
246   [0, 0, -3, 0, -1, 0, 0, 0, 0]            -0.444444  
247  [1, -1, -1, 0, -1, 0, 0, 0, 0]            -0.222222  
248     [0, 0, 1, 0, 2, 0, 1, 0, 0]             0.444444  
249     [0, 0, 0, 0, 0, 0, 0, 0, 0]             0.000000  

[250 rows x 4 columns]

In [202]:
sum(avg_improvement)/len(avg_improvement)

0.09422222222222223

In [233]:
improvement_ct = 0
for i in range(len(result_2018)):
   if result_2018['Average Improvement'][i] > 0:
       improvement_ct += 1
improvement_ct/len(result_2018)

0.428

The average improvement in subjects for 2018 was 0.0942.

In [206]:
study_2019 = pd.DataFrame(study_2019).reset_index()
result_2019 = study_2019.groupby('base_id').apply(create_virus_dict_cross).reset_index(name='Virus Measurements')
improvement_arr = []
for i in range(len(result_2019['Virus Measurements'])):
    temp_improvement_arr = []
    for virus in result_2019['Virus Measurements'][i]:
        delta = result_2019['Virus Measurements'][i][virus]
        temp_improvement_arr.append(delta[len(delta)-1]-delta[0])
    improvement_arr.append(temp_improvement_arr)

result_2019['Improvement'] = improvement_arr
avg_improvement = []
for i in range(len(result_2019['Improvement'])):
    avg_improvement.append(sum(result_2019['Improvement'][i]) / len(result_2019['Improvement'][i]))

result_2019['Average Improvement'] = avg_improvement

C:\Users\anall\AppData\Local\Temp\ipykernel_35012\3916131061.py:2: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  result_2019 = study_2019.groupby('base_id').apply(create_virus_dict_cross).reset_index(name='Virus Measurements')


In [207]:
result_2019

base_id                                 Virus Measurements  \
0       001  {'H1N1 A/Brisbane/2/2018': [3, 4, 3], 'H1N1 A/...   
1      004C  {'H1N1 A/Brisbane/2/2018': [5, 6, 6], 'H1N1 A/...   
2       005  {'H1N1 A/Brisbane/2/2018': [1, 1, 0], 'H1N1 A/...   
3      006C  {'H1N1 A/Brisbane/2/2018': [4, 5], 'H1N1 A/Bri...   
4      007C  {'H1N1 A/Brisbane/2/2018': [2, 2], 'H1N1 A/Bri...   
..      ...                                                ...   
456     528  {'H1N1 A/Brisbane/2/2018': [4, 5], 'H1N1 A/Bri...   
457     529  {'H1N1 A/Brisbane/2/2018': [3, 4], 'H1N1 A/Bri...   
458     530  {'H1N1 A/Brisbane/2/2018': [0, 5], 'H1N1 A/Bri...   
459     531  {'H1N1 A/Brisbane/2/2018': [2, 2, 4], 'H1N1 A/...   
460     535  {'H1N1 A/Brisbane/2/2018': [1, 6], 'H1N1 A/Bri...   

              Improvement  Average Improvement  
0      [0, 5, 0, 1, 1, 0]             1.166667  
1     [1, 0, 1, 0, 0, -1]             0.166667  
2    [-1, 0, 0, 0, 0, -1]            -0.333333  
3      [1, 2, 1, 0, 1, 0]             0.833333  
4      [0, 0, 0, 0, 0, 0]             0.000000  
..                    ...                  ...  
456    [1, 0, 1, 0, 0, 0]             0.333333  
457   [1, -1, 1, 0, 0, 2]             0.500000  
458    [5, 1, 6, 0, 4, 7]             3.833333  
459    [2, 1, 3, 0, 2, 1]             1.500000  
460    [5, 0, 1, 0, 5, 4]             2.500000  

[461 rows x 4 columns]

In [211]:
sum(avg_improvement)/len(avg_improvement)

0.5336225596529284

In [235]:
improvement_ct = 0
for i in range(len(result_2019)):
   if result_2019['Average Improvement'][i] > 0:
       improvement_ct += 1
improvement_ct/len(result_2019)

0.6203904555314533

The average improvement in subjects in 2019 was 0.5336. 

In [213]:
study_2020 = pd.DataFrame(study_2020).reset_index()
result_2020 = study_2020.groupby('base_id').apply(create_virus_dict_cross).reset_index(name='Virus Measurements')
improvement_arr = []
for i in range(len(result_2020['Virus Measurements'])):
    temp_improvement_arr = []
    for virus in result_2020['Virus Measurements'][i]:
        delta = result_2020['Virus Measurements'][i][virus]
        temp_improvement_arr.append(delta[len(delta)-1]-delta[0])
    improvement_arr.append(temp_improvement_arr)

result_2020['Improvement'] = improvement_arr
avg_improvement = []
for i in range(len(result_2020['Improvement'])):
    avg_improvement.append(sum(result_2020['Improvement'][i]) / len(result_2020['Improvement'][i]))

result_2020['Average Improvement'] = avg_improvement

C:\Users\anall\AppData\Local\Temp\ipykernel_35012\3305052475.py:2: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  result_2020 = study_2020.groupby('base_id').apply(create_virus_dict_cross).reset_index(name='Virus Measurements')


In [218]:
result_2020

base_id                                 Virus Measurements  \
0       001  {'H1N1 A/Brisbane/2/2018': [3, 3, 3], 'H1N1 A/...   
1      004C  {'H1N1 A/Brisbane/2/2018': [6, 7], 'H1N1 A/Cal...   
2       005  {'H1N1 A/Brisbane/2/2018': [0, 1, 0], 'H1N1 A/...   
3       008  {'H1N1 A/Brisbane/2/2018': [2, 4, 3], 'H1N1 A/...   
4      009C  {'H1N1 A/Brisbane/2/2018': [1, 5, 5], 'H1N1 A/...   
..      ...                                                ...   
334     539  {'H1N1 A/Brisbane/2/2018': [0, 5, 5], 'H1N1 A/...   
335     540  {'H1N1 A/Brisbane/2/2018': [3, 5, 4], 'H1N1 A/...   
336     542  {'H1N1 A/Brisbane/2/2018': [7, 7], 'H1N1 A/Cal...   
337     549  {'H1N1 A/Brisbane/2/2018': [5, 6], 'H1N1 A/Cal...   
338     555  {'H1N1 A/Brisbane/2/2018': [5, 5], 'H1N1 A/Cal...   

          Improvement  Average Improvement  
0    [0, 0, -1, 0, 0]                 -0.2  
1     [1, 0, 1, 1, 2]                  1.0  
2    [0, 0, 2, 1, -2]                  0.2  
3     [1, 0, 0, 1, 1]                  0.6  
4     [4, 2, 4, 3, 2]                  3.0  
..                ...                  ...  
334   [5, 1, 4, 2, 1]                  2.6  
335   [1, 1, 2, 3, 0]                  1.4  
336   [0, 0, 0, 1, 1]                  0.4  
337   [1, 0, 0, 1, 1]                  0.6  
338   [0, 1, 1, 0, 1]                  0.6  

[339 rows x 4 columns]

In [220]:
sum(avg_improvement)/len(avg_improvement)

0.3952802359882006

In [237]:
improvement_ct = 0
for i in range(len(result_2020)):
   if result_2020['Average Improvement'][i] > 0:
       improvement_ct += 1
improvement_ct/len(result_2020)

0.6253687315634219

The average improvement in subjects in 2020 was 0.3952.

In [223]:
study_2021 = pd.DataFrame(study_2021).reset_index()
result_2021 = study_2021.groupby('base_id').apply(create_virus_dict_cross).reset_index(name='Virus Measurements')
improvement_arr = []
for i in range(len(result_2021['Virus Measurements'])):
    temp_improvement_arr = []
    for virus in result_2021['Virus Measurements'][i]:
        delta = result_2021['Virus Measurements'][i][virus]
        temp_improvement_arr.append(delta[len(delta)-1]-delta[0])
    improvement_arr.append(temp_improvement_arr)

result_2021['Improvement'] = improvement_arr
avg_improvement = []
for i in range(len(result_2021['Improvement'])):
    avg_improvement.append(sum(result_2021['Improvement'][i]) / len(result_2021['Improvement'][i]))

result_2021['Average Improvement'] = avg_improvement

C:\Users\anall\AppData\Local\Temp\ipykernel_35012\1961319788.py:2: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  result_2021 = study_2021.groupby('base_id').apply(create_virus_dict_cross).reset_index(name='Virus Measurements')


In [224]:
result_2021

base_id                                 Virus Measurements  \
0       001  {'H1N1 A/Brisbane/2/2018': [3, 7], 'H1N1 A/Cal...   
1       005  {'H1N1 A/Brisbane/2/2018': [0, 2], 'H1N1 A/Cal...   
2       008  {'H1N1 A/Brisbane/2/2018': [3, 5, 4], 'H1N1 A/...   
3      009C  {'H1N1 A/Brisbane/2/2018': [5, 5], 'H1N1 A/Cal...   
4       011  {'H1N1 A/Brisbane/2/2018': [6, 7], 'H1N1 A/Cal...   
..      ...                                                ...   
332     560  {'H1N1 A/Brisbane/2/2018': [4, 5], 'H1N1 A/Cal...   
333     561  {'H1N1 A/Brisbane/2/2018': [6, 6, 6], 'H1N1 A/...   
334     562  {'H1N1 A/Brisbane/2/2018': [4, 5], 'H1N1 A/Cal...   
335     563  {'H1N1 A/Brisbane/2/2018': [3, 4], 'H1N1 A/Cal...   
336     564  {'H1N1 A/Brisbane/2/2018': [4, 5], 'H1N1 A/Cal...   

        Improvement  Average Improvement  
0      [4, 4, 4, 5]                 4.25  
1     [2, 1, -1, 2]                 1.00  
2      [1, 2, 0, 1]                 1.00  
3    [0, -1, -1, 0]                -0.50  
4      [1, 3, 2, 3]                 2.25  
..              ...                  ...  
332    [1, 1, 1, 3]                 1.50  
333    [0, 1, 0, 1]                 0.50  
334    [1, 2, 1, 2]                 1.50  
335    [1, 1, 1, 1]                 1.00  
336    [1, 2, 1, 3]                 1.75  

[337 rows x 4 columns]

In [227]:
sum(avg_improvement)/len(avg_improvement)

1.072700296735905

In [239]:
improvement_ct = 0
for i in range(len(result_2021)):
   if result_2021['Average Improvement'][i] > 0:
       improvement_ct += 1
improvement_ct/len(result_2021)

0.8100890207715133

In [241]:
improvement_arr = [0.2951858108108108, 0.21309963099630996, 0.09422222222222223, 0.5336225596529284, 0.3952802359882006, 1.072700296735905]
percentage_arr = [0.6351351351351351*100, 0.5018450184501845*100, 0.428*100, 0.6203904555314533*100, 0.6253687315634219*100, 0.8100890207715133*100]

In [243]:
study_table = pd.DataFrame(improvement_arr, columns=['Average Improvement'])
study_table['Percentage of Improvement(%)'] = percentage_arr
study_table.index = ['2016 UGA', '2017 UGA', '2018 UGA', '2019 UGA', '2020 UGA', '2021 UGA']
study_table

Average Improvement  Percentage of Improvement(%)
2016 UGA             0.295186                     63.513514
2017 UGA             0.213100                     50.184502
2018 UGA             0.094222                     42.800000
2019 UGA             0.533623                     62.039046
2020 UGA             0.395280                     62.536873
2021 UGA             1.072700                     81.008902

2021 vaccine has the best cross reactivity and 2018 has the worst.